flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionAttemptID.java

```diff
@@ -68,19 +117,32 @@ public boolean equals(Object obj) {
             return true;
         } else if (obj != null && obj.getClass() == getClass()) {
             ExecutionAttemptID that = (ExecutionAttemptID) obj;
-            return that.executionAttemptId.equals(this.executionAttemptId);
+            return that.executionGraphId.equals(this.executionGraphId)
+                    && that.executionVertexId.equals(this.executionVertexId)
+                    && that.attemptNumber == this.attemptNumber;
         } else {
             return false;
         }
     }
 
     @Override
     public int hashCode() {
-        return executionAttemptId.hashCode();
+        return Objects.hash(executionGraphId, executionVertexId, attemptNumber);
     }
 
     @Override
     public String toString() {
-        return executionAttemptId.toString();
+        return String.format(
+                "%s_%s_%d", executionGraphId.toString(), executionVertexId, attemptNumber);
+    }
+
+    public String getLogString() {
+        if (DefaultExecutionGraph.LOG.isDebugEnabled()) {
+            return toString();
+        } else {
+            return String.format(
+                    "%s_%s_%d",
+                    executionGraphId.toString().substring(0, 4), executionVertexId, attemptNumber);
```

Comment:
logging the EG ID is fine, but this can't be done in isolation. There must be a way to correlate this id with a specific job submission; for example we could log the EG ID when the EG for a given job was created.

Reply:
It might also make sense to return a more structured representation that actually tells the reader what they are looking at.

Reply:
> logging the EG ID is fine, but this can't be done in isolation. There must be a way to correlate this id with a specific job submission; for example we could log the EG ID when the EG for a given job was created.

Totally agree. Will add a log when creating the EG.

Reply:
> It might also make sense to return a more structured representation that actually tells the reader what they are looking at.

Agreed. Actually I'm planning to open a separate JIRA&pr to refine the logs which contains an `ExecutionAttemptID`.
Currently, an execution is usually represented as "`job vertex name` (`subtaskIndex+1`/`vertex parallelism`) (`attemptId`)", which may be redundant after this refactoring work. I'm planning to change the format to be "`job vertex name` (`short ExecutionGraphID`:`JobVertexID`) (`subtaskIndex+1`/`vertex parallelism`) (`#attemptNumber`)" and avoid directly display the `ExecutionAttemptID`. This displayed `JobVertexID` can also help to distinguish job vertices of the same name, which is common in DataStream jobs (e.g. multiple `Map`).

The logs are spread among multiple classes and needs some further examination. Therefore I'd like to remove the current `ExecutionAttemptID#getLogString()` and do this work in a separate task.

WDYT?

Reply:
> Can we put the work of introducing EG ID into a separate commit?

Sure. A separate commit "[Introduce ExecutionGraph](https://github.com/apache/flink/pull/19747/commits/cb37f0a552595b35ce4e64e9233eabefc618fcb4)" is added.

Reply:
The log to print the ExecutionGraphID is added.

Reply:
> The logs are spread among multiple classes and needs some further examination. Therefore I'd like to remove the current ExecutionAttemptID#getLogString() and do this work in a separate task.

Sure, this sounds fine.

Reply:
FLINK-27710 is created to refine the representation of Execution.